src/dawn/SIR/SIR.cpp

```diff
@@ -174,4 +174,137 @@ std::shared_ptr<sir::StencilCall> sir::StencilCall::clone() const {
   return call;
 }
 
+bool SIR::operator==(const SIR& rhs) const {
+  bool retval = true;
+
+  if((Stencils.size() == rhs.Stencils.size()) &&
+     (StencilFunctions.size() == rhs.StencilFunctions.size())) {
+
+    if(Stencils.size() > 0) {
+      retval &= std::equal(Stencils.begin(), Stencils.end(), rhs.Stencils.begin(),
+                           sharedpointer_comparison<sir::Stencil>);
+    }
+
+    if(StencilFunctions.size() > 0) {
+      retval &=
+          std::equal(StencilFunctions.begin(), StencilFunctions.end(), rhs.StencilFunctions.begin(),
+                     sharedpointer_comparison<sir::StencilFunction>);
+    }
+
+    if(GlobalVariableMap != nullptr && rhs.GlobalVariableMap != nullptr) {
+      if(GlobalVariableMap->size() == rhs.GlobalVariableMap->size()) {
+        if(GlobalVariableMap->size() > 0) {
+          retval &=
+              pointerMapComparison(*(GlobalVariableMap.get()), *(rhs.GlobalVariableMap.get()));
+        }
+      } else {
+        return false;
+      }
+    } else {
+      return GlobalVariableMap == rhs.GlobalVariableMap;
+    }
+
+    return retval;
+  }
+  return false;
+}
+bool SIR::operator!=(const SIR& rhs) const { return !(*this == rhs); }
+
+bool sir::Stencil::operator==(const sir::Stencil& rhs) const {
+  bool retval = true;
+  if(Fields.size() != rhs.Fields.size()) {
+    return false;
+  }
+  retval &= (Name == rhs.Name);
```

Comment:
The problem with using
```
retval &= ...
```
is that even *if* it is false, you don't exit the function but do the whole comparison. Maybe it is more efficient to just always return false as soon as you see a mismatch and if you pass all the checks simply return `true`

Reply:
I was thinking about this and am not sure what is better. If we expect SIRs to match most of the time (which I think they do), we do not want early exits to better pipeline. 
Of course we lose performance for mismatches. How certain are you that early exits are worth it?

Reply:
I wouldn't worry much about piplining at this *high* level (this should be done by the compiler anyway - and actually is done). Early exits will definitely improve performance drastically and will not have a noticeable difference if the SIR are equal. See https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code